compiler/src/iree/compiler/GlobalOptimization/DetachElementwiseFromNamedOps.cpp

```diff
@@ -63,12 +63,35 @@ struct DetachElementwisePattern
       return failure();
     }
     auto outputType = llvm::cast<RankedTensorType>(outputOperand.getType());
+    int64_t outputRank = outputType.getRank();
     if (!outputType.getElementType().isIntOrFloat())
       return failure();
     auto elementType = outputType.getElementType();
 
     Location loc = linalgOp.getLoc();
 
+    // verify the original output affine map is parallel
+    auto outputMap = mlir::compressUnusedDims(
+        linalgOp.getMatchingIndexingMap(outputOperands.front()));
+    SmallVector<utils::IteratorType> iterators;
+    iterators.reserve(outputMap.getNumResults());
+    for (int i = 0, e = outputMap.getNumResults(); i < e; ++i) {
+      auto expr = dyn_cast<AffineDimExpr>(outputMap.getResult(i));
+      if (!expr)
+        return rewriter.notifyMatchFailure(
+            linalgOp, "output affine map has a non dim expression at " +
+                          std::to_string(i));
+      int pos = expr.getPosition();
+      auto attr = linalgOp.getIteratorTypesArray()[pos];
+      if (!linalg::isParallelIterator(attr))
+        return rewriter.notifyMatchFailure(
+            linalgOp, "output iterator type is not parallel at position " +
+                          std::to_string(pos));
+      iterators.push_back(attr);
+    }
+
```

Comment:
Is it even possible for a reduction iterator to be used on an output (especially since it must be a conv or contraction)? If not, this could just be simplified to something like:


```cpp
if(!outputMap.isProjectedPermutation())
  	return failure()
SmallVector<utils::IteratorType> iterators(outputRank, utils::IteratorType::parallel)
```

Reply:
This is much cleaner, thanks. 

Although, I'm thinking that we probably don't need to check anything on the output tensor indexing maps. The logic here is really simple: whatever terms would get accumulated to an init value instead accumulates to a zero init value, then we add back the original init values. 

To be fair, I don't know enough about what is actually done with the iterator types to know what is valid here. 

1. Is it possible for there to be a parallel iterator that is *not* represented in the output tensor indexing maps? 
2. Is it possible that an output tensor indexing map has some affine term with a reduction iterator?
3. If the answer to the following two questions is no, then is there a reason the iterator types even need to be specified when building a linalg op? That is, why can't these iterator types be inferred from the maps themselves?